src/userobjects/MoabSkinner.C

```diff
@@ -0,0 +1,1008 @@
+  _moab->tag_set_data(tag, &ent, 1, namebuf);
+  delete[] namebuf;
+}
+
+void
+MoabSkinner::setTagData(moab::Tag tag, moab::EntityHandle ent, void * data)
+{
+  _moab->tag_set_data(tag, &ent, 1, data);
+}
+
+unsigned int
+MoabSkinner::nBins() const
+{
+  return _n_block_bins * _n_density_bins * _n_temperature_bins;
+}
+
+void
+MoabSkinner::sortElemsByResults()
+{
+  _elem_bins.clear();
+  _elem_bins.resize(nBins());
+
+  // accumulate information for printing diagnostics
+  std::vector<unsigned int> n_block_hits(_n_block_bins, 0);
+  std::vector<unsigned int> n_temp_hits(_n_temperature_bins, 0);
+  std::vector<unsigned int> n_density_hits(_n_density_bins, 0);
+
+  for (unsigned int e = 0; e < _fe_problem.mesh().nElem(); ++e)
+  {
+    const Elem * const elem = _fe_problem.mesh().queryElemPtr(e);
+    if (!elem)
+      continue;
+
+    Point p = elem->vertex_average();
+
+    // bin by subdomain ID
+    auto iMat = getSubdomainBin(elem);
+    n_block_hits[iMat] += 1;
+
+    // bin by density
+    auto iDenBin = getDensityBin(elem);
+    n_density_hits[iDenBin] += 1;
+
+    // bin by temperature
+    auto iBin = getTemperatureBin(elem);
+    n_temp_hits[iBin] += 1;
+
+    // Sort elem into a bin
+    auto iSortBin = getBin(iBin, iDenBin, iMat);
+    _elem_bins.at(iSortBin).insert(elem->id());
+  }
+
+  VariadicTable<unsigned int, std::string, unsigned int> vtt({"Bin", "Range (K)", "# Elems"});
+  VariadicTable<unsigned int, std::string, unsigned int> vtd({"Bin", "Range (kg/m3)", "# Elems"});
+
+  for (unsigned int i = 0; i < _n_temperature_bins; ++i)
+    vtt.addRow(i,
+               std::to_string(_temperature_bin_bounds[i]) + " to " +
+                   std::to_string(_temperature_bin_bounds[i + 1]),
+               n_temp_hits[i]);
+
+  for (unsigned int i = 0; i < _n_density_bins; ++i)
+    vtd.addRow(i,
+               std::to_string(_density_bin_bounds[i]) + " to " +
+                   std::to_string(_density_bin_bounds[i + 1]),
+               n_density_hits[i]);
+
+  if (_verbose)
+  {
+    _console << "Mapping of Elements to Temperature Bins:" << std::endl;
+    vtt.print(_console);
+
+    if (_bin_by_density)
+    {
+      _console << "\nMapping of Elements to Density Bins:" << std::endl;
+      vtd.print(_console);
+    }
+  }
+}
+
+unsigned int
+MoabSkinner::getTemperatureBin(const Elem * const elem) const
+{
+  auto dof = elem->dof_number(_fe_problem.getAuxiliarySystem().number(), _temperature_var_num, 0);
+  auto value = (*_serialized_solution)(dof);
+
+  // TODO: add option to truncate instead
+  if (value < _temperature_min)
+    mooseError("Variable '",
+               _temperature_name,
+               "' has value below minimum range of bins. "
+               "Please decrease 'temperature_min'.\n\n"
+               "  value: ",
+               value,
+               "\n  temperature_min: ",
+               _temperature_min);
+
+  if (value > _temperature_max)
+    mooseError("Variable '",
```

Comment:
Yeah, I'd like to have an option to truncate in the future. I think I have a similar to-do item in OpenMC (in case you went like 1K over the loaded XS data sets). I think for now, I'd like to keep it as an error, and add an option like `truncate = true` to throw a warning instead (in a future PR). What do you think?

Reply:
Something like this may have already been added in OpenMC for temperature interpolation: https://github.com/openmc-dev/openmc/pull/2265

I'm game to bump this to a future PR for now. Maybe create an issue and link it here so we don't lose track?

Reply:
Oh nice, we can definitely take advantage of that in Cardinal. I'll make an issue to keep track of it.

Reply:
https://github.com/neams-th-coe/cardinal/issues/234